cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -475,77 +583,88 @@ func getAPIGroupPreferredVersion(client discovery.DiscoveryInterface, APIGroup s
 }
 
 func (c *machineController) scalableResourceProviderIDs(scalableResource *unstructured.Unstructured) ([]string, error) {
-	machines, err := c.listMachinesForScalableResource(scalableResource)
-	if err != nil {
-		return nil, fmt.Errorf("error listing machines: %v", err)
-	}
-
 	var providerIDs []string
-	for _, machine := range machines {
-		providerID, found, err := unstructured.NestedString(machine.UnstructuredContent(), "spec", "providerID")
+	if scalableResource.GetKind() == machinePoolKind {
+		providerIDList, found, err := unstructured.NestedStringSlice(scalableResource.UnstructuredContent(), "spec", "providerIDList")
 		if err != nil {
 			return nil, err
 		}
+		if !found {
+			klog.Warningf("Machine Pool %q has no providerIDList", scalableResource.GetName())
+		} else {
+			providerIDs = providerIDList
+		}
```

Comment:
nit: might be clearer to read by returning earlier here and avoid having the whole main logic within an else block